File: frontend/src/services/bankSync.ts
import { request } from "./api";

export interface BankAccount {
  id: string;
  connection_id: string;
  external_account_id: string;
  account_name: string;
  account_type: string;
  account_subtype?: string;
  mask?: string;
  currency: string;
  current_balance?: number;
  available_balance?: number;
  is_active: boolean;
  created_at: string;
}

export interface BankConnection {
  id: string;
  provider: string;
  institution_id: string;
  institution_name: string;
  status: "connected" | "syncing" | "error" | "disconnected";
  last_sync_at?: string;
  sync_status: "idle" | "syncing" | "success" | "failed";
  sync_error_message?: string;
  created_at: string;
  accounts: BankAccount[];
}

export interface BankConnectRequest {
  provider?: string;
  institution_id?: string;
  auth_code?: string;
  state?: string;
}

export interface BankSyncResult {
  connection_id: string;
  status: string;
  imported_count: number;
  duplicates_skipped: number;
  accounts_synced: number;
  synced_at: string;
  message: string;
}

export interface ProviderInstitution {
  institution_id: string;
  institution_name: string;
  logo_url?: string;
  supported_features: string[];
  is_mock?: boolean;
}

export interface ProviderStatus {
  provider: string;
  is_mock: boolean;
  is_configured: boolean;
  supported_institutions: ProviderInstitution[];
}

export const bankSyncApi = {
  getConnections: (): Promise<BankConnection[]> => {
    return request("/api/bank-connections");
  },

  getProviderStatus: (): Promise<ProviderStatus> => {
    return request("/api/bank-connections/providers");
  },

  connectBank: (data: BankConnectRequest): Promise<BankConnection> => {
    return request("/api/bank-connections/connect", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },

  getAccounts: (connectionId: string): Promise<BankAccount[]> => {
    return request(`/api/bank-connections/${connectionId}/accounts`);
  },

  triggerSync: (connectionId: string): Promise<BankSyncResult> => {
    return request(`/api/bank-connections/${connectionId}/sync`, {
      method: "POST",
    });
  },

  disconnectBank: (connectionId: string): Promise<void> => {
    return request(`/api/bank-connections/${connectionId}`, {
      method: "DELETE",
    });
  },
};

File: backend/schemas/bank_sync.py
"""
Pydantic Schemas for Bank Synchronization & Open Banking (Task 19)
Enforces token confidentiality: provider secrets are strictly omitted from all client responses.
"""

from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, ConfigDict


class BankAccountResponse(BaseModel):
    id: str
    connection_id: str
    external_account_id: str
    account_name: str
    account_type: str
    account_subtype: Optional[str] = None
    mask: Optional[str] = None
    currency: str = "INR"
    current_balance: Optional[float] = None
    available_balance: Optional[float] = None
    is_active: bool = True
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class BankConnectionResponse(BaseModel):
    id: str
    provider: str
    institution_id: str
    institution_name: str
    status: str
    last_sync_at: Optional[datetime] = None
    sync_status: str
    sync_error_message: Optional[str] = None
    created_at: datetime
    accounts: List[BankAccountResponse] = []

    model_config = ConfigDict(from_attributes=True)


class BankConnectRequest(BaseModel):
    provider: Optional[str] = "mock"
    institution_id: Optional[str] = None
    auth_code: Optional[str] = None
    state: Optional[str] = None


class BankSyncResponse(BaseModel):
    connection_id: str
    status: str
    imported_count: int
    duplicates_skipped: int
    accounts_synced: int
    synced_at: datetime
    message: str


class ProviderStatusResponse(BaseModel):
    provider: str
    is_mock: bool
    is_configured: bool
    supported_institutions: List[Dict[str, Any]]
